bufferevent_openssl.c

```diff
@@ -259,7 +259,9 @@ conn_closed(struct bufferevent_ssl *bev_ssl, int when, int errcode, int ret)
 		bufferevent_ssl_put_error(bev_ssl, errcode);
 		break;
 	case SSL_ERROR_SSL:
-		/* Protocol error. */
+		/* Protocol error; possibly a dirty shutdown. */
```

Comment:
I started looking at this independently before I found this pull request, and came up with a slightly different formulation to fix this. 
SSL_ERROR_SSL is not guaranteed to indicate that a system call has failed, so looking at "ret" does not reliably indicate that a read call has given an EOF -  it's possibly better to look at the error code here to ensure that it's caused by an EOF on a read:

```
diff --git a/bufferevent_openssl.c b/bufferevent_openssl.c
index 6ace1e3a..5bdda402 100644
--- a/bufferevent_openssl.c
+++ b/bufferevent_openssl.c
@@ -259,6 +259,18 @@ conn_closed(struct bufferevent_ssl *bev_ssl, int when, int errcode, int ret)
                bufferevent_ssl_put_error(bev_ssl, errcode);
                break;
        case SSL_ERROR_SSL:
+#ifdef SSL_R_UNEXPECTED_EOF_WHILE_READING
+               err = ERR_peek_error();
+               if (ERR_GET_LIB(err) == ERR_LIB_SSL &&
+                   ERR_GET_REASON(err) == SSL_R_UNEXPECTED_EOF_WHILE_READING) {
+                       /* Possibly a clean shutdown. */
+                       if (SSL_get_shutdown(bev_ssl->ssl) & SSL_RECEIVED_SHUTDOWN)
+                               event = BEV_EVENT_EOF;
+                       else
+                               dirty_shutdown = 1;
+                       break;
+               }
+#endif
```

Reply:
This looks better, will apply your hunk, thanks!

Reply:
By the way, @peadar maybe we should check here for `SSL_R_TLSV1_ALERT_DECODE_ERROR`, as suggested by @loqs here https://github.com/libevent/libevent/issues/1250#issuecomment-1097249859, what do you think?

Reply:
Also what I don't understand as of now, why the test does not fails in my env.
Compiled libevent with upstream openssl openssl/openssl@12e488367d